pkg/ddevapp/ddevapp.go

```diff
@@ -1654,6 +1654,20 @@ Fix with 'ddev config global --required-docker-compose-version="" --use-docker-c
 	output.UserOut.Printf("Waiting for containers to become ready: %v", dependers)
 	waitErr := app.Wait(dependers)
 
+	if app.Database.Type == nodeps.MySQL && (app.Database.Version == nodeps.MySQL80 || app.Database.Version == nodeps.MySQL84) && slices.Contains([]string{nodeps.PHP73, nodeps.PHP72, nodeps.PHP71, nodeps.PHP70, nodeps.PHP56}, app.PHPVersion) {
+		alterString := `ALTER USER 'db'@'%' IDENTIFIED WITH mysql_native_password BY 'db';
+			ALTER USER 'root'@'%' IDENTIFIED WITH mysql_native_password BY 'root';
+			ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'root';`
+		_, _, err = app.Exec(&ExecOpts{
+			Cmd:     fmt.Sprintf(`mysql -uroot -proot -e "%s"`, alterString),
+			Service: `db`,
+		})
+		if err != nil {
+			util.Warning("unable to set mysql_native_password db password: %v", err)
+		}
+		util.Debug(`mysql 8, php 5-7.3, set mysql_native_password`)
+	}
+
```

Comment:
From https://github.com/ddev/ddev/pull/6589#discussion_r1886571895

(The code is not complete, because it's not easy to write here in suggestions.)

```suggestion
	if !slices.Contains(app.OmitContainers, "db") && app.Database.Type == nodeps.MySQL && (app.Database.Version == nodeps.MySQL80 || app.Database.Version == nodeps.MySQL84) && slices.Contains([]string{nodeps.PHP73, nodeps.PHP72, nodeps.PHP71, nodeps.PHP70, nodeps.PHP56}, app.PHPVersion) {
		// if mysql8.0, run app.Exec for:
		// `mysql -uroot -proot -e 'SELECT @@default_authentication_plugin;' -Ns 2>/dev/null` == "mysql_native_password"
		// else if mysql8.4, run app.Exec for:
		// `mysql -uroot -proot -e 'SELECT PLUGIN_STATUS FROM INFORMATION_SCHEMA.PLUGINS WHERE PLUGIN_NAME="mysql_native_password";' -Ns 2>/dev/null` == "ACTIVE"
		// if one of the condition above is true, run this alterString:
		alterString := `ALTER USER 'db'@'%' IDENTIFIED WITH mysql_native_password BY 'db';
			ALTER USER 'db'@'localhost' IDENTIFIED WITH mysql_native_password BY 'db';
			ALTER USER 'root'@'%' IDENTIFIED WITH mysql_native_password BY 'root';
			ALTER USER 'root'@'localhost' IDENTIFIED WITH mysql_native_password BY 'root';`
		_, _, err = app.Exec(&ExecOpts{
			Cmd:     fmt.Sprintf(`mysql -uroot -proot -e "%s"`, alterString),
			Service: `db`,
		})
		if err != nil {
			util.Warning("unable to set mysql_native_password db password: %v", err)
		}
		util.Debug(`mysql 8, php 5-7.3, set mysql_native_password`)
	}

```